johnnyfive/gmail.py

```diff
@@ -0,0 +1,396 @@
+    fromaddr : `str`, optional
+        Sender email address [Default: Value from [gmailSetup]]
+    """
+    def __init__(self, toaddr, subject, message_text, fromname=None,
+                 fromaddr=None, interactive=False):
+        # Load default `fromaddr`` if None passed in
+        if not fromaddr:
+            fromaddr = utils.read_ligmos_conffiles('gmailSetup').user
+
+        # Initialize the Gmail connection
+        self.service = setup_gmail(interactive=interactive)
+
+        # Build the container for a multipart MIME message
+        self.message = multipart.MIMEMultipart()
+        self.message['to'] = toaddr if isinstance(toaddr, str) else ','.join(toaddr)
+        self.message['from'] = f"{fromname} <{fromaddr}>" if fromname else fromaddr
+        self.message['subject'] = subject
+
+        # Place the text into the message
+        self.message.attach( text.MIMEText(message_text))
+
+    def add_attachment(self, file):
+        """add_attachment _summary_
+
+        _extended_summary_
+
+        Parameters
+        ----------
+        file : `str`
+            Filename of the attachment
+        """
+        # For the attachment, guess the MIME type for reading it in
+        content_type, encoding = mimetypes.guess_type(file)
+
+        # Set unknown type
+        if content_type is None or encoding is not None:
+            content_type = 'application/octet-stream'
+
+        # Case out the content type
+        main_type, sub_type = content_type.split('/', 1)
+        if main_type == 'text':
+            with open(file, 'rb') as fp:
+                attachment = text.MIMEText(fp.read(), _subtype=sub_type)
+        elif main_type == 'image':
+            with open(file, 'rb') as fp:
+                attachment = image.MIMEImage(fp.read(), _subtype=sub_type)
+        elif main_type == 'audio':
+            with open(file, 'rb') as fp:
+                attachment = audio.MIMEAudio(fp.read(), _subtype=sub_type)
+        else:
+            with open(file, 'rb') as fp:
+                attachment = base.MIMEBase(main_type, sub_type)
+                attachment.set_payload(fp.read())
+
+        # Add the attachment to the email message
+        attachment.add_header('Content-Disposition', 'attachment',
+                           filename=os.path.basename(file))
+        self.message.attach(attachment)
+
+    def send(self):
+        """send Send the GmailMessage
+
+        _extended_summary_
+
+        Returns
+        -------
+        `dict`
+            The sent message object
+        """
+        # Take the message object, and 64-bit encode it properly for sending
+        encoded_message = base64.urlsafe_b64encode(self.message.as_bytes())
+        # The sendable message is a dictionary containing the raw decoded thing
+        sendable_message = {'raw': encoded_message.decode()}
+
+        # Go into a possibly infinite loop!
+        # TODO: Make this NOT an infinite loop
+        while not self.service:
```

Comment:
```suggestion
        nTries = 5
        thisTries = 0
        if self.service is None and thisTries > nTries:
```

Reply:
You'll have to increment thisTries in the resulting if loop but that's not in the suggestion I just realized.

Reply:
Removed infinite loop, left plenty of opportunities to retry to connect with Gmail.